File: simulation/estimaters.py
import numpy as np
from dataclasses import dataclass
import sys
from pathlib import Path
sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
import windows


@dataclass
class EstimateData:
    freq_estimated: float
    amplitude_estimated: float
    phase_estimated: float = 0.0

def _no_interp(#没有插值算法,模拟不加插值算法的情形
        spectrum: np.ndarray, #复数谱
        peak_bin: int,
        fs_hz: float,
        n_pts: int,
        coherent_gain:float
        ) -> EstimateData:
    if peak_bin > n_pts // 2:
        peak_bin = (n_pts // 2) - peak_bin
    output_EstimateData = EstimateData(
            freq_estimated=peak_bin * (fs_hz / n_pts),
            amplitude_estimated=np.abs(spectrum[peak_bin]) / (n_pts * coherent_gain),
            phase_estimated=np.angle(spectrum[peak_bin])
            )
    return output_EstimateData

def _estimate_phase(
        angle: float, #该点辐角
        freq_delta: float, #频差δ
        n_pts: int #fft点数
        ):
    return angle - np.pi * freq_delta * (n_pts - 1) / n_pts

def _parabolic_interp( #抛物线插值法
                    spectrum: np.ndarray, #复数谱
                     peak_bin: int, #峰值bin
                     fs_hz: float,
                     n_pts: int,
                     coherent_gain: float #相干增益
                     ) -> EstimateData:
    y_minus1 = spectrum[peak_bin - 1]
    y_minus1 = np.abs(y_minus1)
    y_0 = spectrum[peak_bin]
    y_0 = np.abs(y_0)
    y_plus1 = spectrum[peak_bin + 1]
    y_plus1 = np.abs(y_plus1)
    freq_delta = 0.5 * (y_minus1 - y_plus1) / (y_minus1 - 2 * y_0 + y_plus1)
    a = (y_minus1 + y_plus1 - 2 * y_0)#抛物线方程中的a
    y_peak = y_0 - a * (freq_delta ** 2)
    if peak_bin > n_pts // 2:
        peak_bin = (n_pts // 2) - peak_bin
    freq_estimated= (peak_bin + freq_delta) * (fs_hz / n_pts)
    amplitude_estimated = y_peak / (n_pts * coherent_gain)
    angle_ori = np.angle(spectrum[peak_bin])
    
    phase_angle = _estimate_phase(
            angle=angle_ori,
            freq_delta=freq_delta,
            n_pts=n_pts
            )

    output_EstimateData = EstimateData(
            freq_estimated=freq_estimated,
            amplitude_estimated=amplitude_estimated,
            phase_estimated=phase_angle
            )
    
    return output_EstimateData
    
def _delta_rect_candan( #candan插值算法
        spectrum_complex: np.ndarray, #candan插值算法需要复数谱
        peak_bin: int, #峰值bin
        ) -> float:
    y_minus1 = spectrum_complex[peak_bin - 1]
    y_0 = spectrum_complex[peak_bin]
    y_plus1 = spectrum_complex[peak_bin + 1]
    freq_delta = np.real((y_minus1 - y_plus1) / (2 * y_0 - y_minus1 -y_plus1))
    return freq_delta

def _get_r_and_alpha( #获得r和alpha(频率修正公式中的常用变量),增加代码复用
        spectrum_complex: np.ndarray,
        peak_bin: int
        ) -> tuple[int,float]:
    y_minus1 = np.abs(spectrum_complex[peak_bin-1])
    y_0 = np.abs(spectrum_complex[peak_bin])
    y_plus1 = np.abs(spectrum_complex[peak_bin+1])
    r = 1 if y_plus1 >= y_minus1 else -1
    alpha = np.abs(spectrum_complex[peak_bin+r]) / y_0
    return r, float(alpha)

#以下为频差估计算法
def _delta_rect_rife(
        spectrum_complex: np.ndarray,
        peak_bin: int
        ) -> float:
    r, alpha = _get_r_and_alpha(spectrum_complex, peak_bin)
    return r * alpha / (1 + alpha)

def _delta_hann_grandke(
        spectrum_complex: np.ndarray,
        peak_bin: int
        ) -> float:
    r, alpha = _get_r_and_alpha(spectrum_complex, peak_bin)
    return r * (2 * alpha - 1) / (alpha + 1)

def _delta_hamming_offelli(
        spectrum_complex: np.ndarray,
        peak_bin: int
        ) -> float:
    r, alpha = _get_r_and_alpha(spectrum_complex, peak_bin)
    return r * (alpha - 0.54) / (0.46 * alpha + 0.54)

def _delta_blackman_andria(
        spectrum_complex: np.ndarray,
        peak_bin: int
        ) -> float:
    r, alpha = _get_r_and_alpha(spectrum_complex, peak_bin)
    return r * (3 * alpha - 1.5) / (alpha + 1.5)

def _delta_blackmanharris_agrez(
        spectrum_complex: np.ndarray,
        peak_bin: int
        ) -> float:
    r, alpha = _get_r_and_alpha(spectrum_complex, peak_bin)
    return r * (4 * alpha - 2) / (alpha + 2)

def _delta_flattop_andria( #注:平顶窗的频率估计不可靠
        spectrum_complex: np.ndarray,
        peak_bin: int
        ) -> float:
    r, alpha = _get_r_and_alpha(spectrum_complex, peak_bin)
    return r * (5.5 * alpha - 4.5) / (alpha + 1)

#以下为幅值估计算法
_sinc_delta = lambda delta: np.pi * delta / np.sin(np.pi * delta)
_m_plus_n_delta_square = lambda m,n,delta: m + n * delta * delta
def _amp_rect(
        x_k0: float,#该频点fft幅值绝对值
        freq_delta: float #频率差δ
        ) -> float:
    return x_k0 * _sinc_delta(freq_delta)

def _amp_hann(
        x_k0: float,
        freq_delta: float
        ) -> float:
    return x_k0 * _sinc_delta(freq_delta) * _m_plus_n_delta_square(1, -1, freq_delta)

def _amp_hamming(
        x_k0: float,
        freq_delta: float
        ) -> float:
    return x_k0 * _sinc_delta(freq_delta) * (_m_plus_n_delta_square(1, -1, freq_delta) / 
                                                 _m_plus_n_delta_square(1, -0.148, freq_delta))

def _amp_blackman(
        x_k0: float,
        freq_delta: float
        ) -> float:
    return x_k0 * _sinc_delta(freq_delta) * (_m_plus_n_delta_square(1, -1, freq_delta) *
                                             _m_plus_n_delta_square(4, -1, freq_delta) /
                                             _m_plus_n_delta_square(4, -1.69, freq_delta))

def _amp_blackmanharris(
        x_k0: float,
        freq_delta: float
        ) -> float:
    return x_k0 * _sinc_delta(freq_delta) * (_m_plus_n_delta_square(1, -1, freq_delta) * 
                                                 _m_plus_n_delta_square(1, -0.25, freq_delta) *
                                                 _m_plus_n_delta_square(1, -0.111, freq_delta))

def _amp_flattop(
        x_k0: float,
        freq_delta: float
        ) -> float:
    return x_k0 * (1 + 0.02 * freq_delta ** 2 + 0.0005 * freq_delta ** 4)

_DELTA_METHODS = {
        "rife": _delta_rect_rife,
        "candan": _delta_rect_candan,
        "granke": _delta_hann_grandke,
        "offelli": _delta_hamming_offelli,
        "andria_blackman": _delta_blackman_andria,
        "agrez": _delta_blackmanharris_agrez,
        "andria_flattop": _delta_flattop_andria
        }

_AMP_METHODS = {
        "rect": _amp_rect,
        "hann": _amp_hann,
        "hamming": _amp_hamming,
        "blackman": _amp_blackman,
        "blackmanharris": _amp_blackmanharris,
        "flattop": _amp_flattop
        }

_AUTO_DELTA_METHODS = { #自动窗函数选择表
        "rect": "candan",
        "hann": "granke",
        "hamming": "offelli",
        "blackman": "andria_blackman",
        "blackmanharris": "agrez",
        "flattop": "andria_flattop"
        }

_WINDOW_CHECK_LIST = { #查看用户是否选择错了窗函数算法
        "rect": ["rife", "candan"],
        "hann": ["granke"],
        "hamming": ["offelli"],
        "blackman": ["andria_blackman"],
        "blackmanharris": ["agrez"],
        "flattop": ["andria_flattop"]
        }

#检查算法是否对应窗函数,防止用户选择错误
def check_algorithm_in_list(window_type: str, algorithm_type: str) -> bool:
    for i in _WINDOW_CHECK_LIST[window_type]:
        if i == algorithm_type:
            return True
    return False

#统一抽象接口
def estimate_freq_amplitude_phase(
        spectrum_complex: np.ndarray, #复数谱
        peak_bin: int, #峰值bin
        fs_hz: float, #采样率
        window_type: str, #窗函数名称
        # rect | hann | hamming | blackman | blackmanharris | flattop
        algorithm_type: str #详情见不同窗函数可适用的频率插值算法表
        ) -> EstimateData:
    n_pts = len(spectrum_complex) #CMSIS-DSP上应该改为 len(spectrum_complex) / 2
    if peak_bin <= 0 or peak_bin > n_pts - 2:
        print("[ERROR] peak bin is too small or too large")
        return EstimateData(
                freq_estimated=float(np.nan),
                amplitude_estimated=float(np.nan),
                phase_estimated=float(np.nan)
                )
    window = windows.generate_window(window_type, n_pts)
    coherent_gain = window.coherent_gain
    #处理auto和之前遗留的抛物线/不加估计算法
    x_k0 = np.abs(spectrum_complex[peak_bin])
    x_k0_angle = np.angle(spectrum_complex[peak_bin])
    method_type = ""
    if algorithm_type == "_no_interp":
        return _no_interp(
                spectrum=spectrum_complex,
                peak_bin=peak_bin,
                fs_hz=fs_hz,
                n_pts=n_pts,
                coherent_gain=coherent_gain
                )
    elif algorithm_type == "parabolic":
        return _parabolic_interp(
                spectrum=spectrum_complex,
                peak_bin=peak_bin,
                fs_hz=fs_hz,
                n_pts=n_pts,
                coherent_gain=coherent_gain
                )
    elif algorithm_type == "auto":
        method_type = _AUTO_DELTA_METHODS[window_type]
    
    elif check_algorithm_in_list(window_type, algorithm_type):
        method_type = algorithm_type
    else:
        print("[ERROR] the algorithm does not fit the windows function")
        return EstimateData(#出错返回nan
                freq_estimated=float(np.nan),
                amplitude_estimated=(np.nan),
                phase_estimated=(np.nan)
                )

    estimated_freq_delta = _DELTA_METHODS[method_type](spectrum_complex, peak_bin)
    if peak_bin > n_pts // 2:
        peak_bin = (n_pts // 2) - peak_bin
    estimated_freq = (peak_bin + estimated_freq_delta) * (fs_hz / n_pts)
    estimated_amplitude = _AMP_METHODS[window_type](x_k0, estimated_freq_delta)
    estimated_amplitude /= n_pts * coherent_gain
    estimated_phase = _estimate_phase(angle=x_k0_angle,
                                        freq_delta=estimated_freq_delta,
                                        n_pts=n_pts)
    return EstimateData(
            freq_estimated=estimated_freq,
            amplitude_estimated=estimated_amplitude,
            phase_estimated=estimated_phase
            )

File: h743_MDK/App/fft/estimaters.c
#include "estimaters.h"
#include <math.h>

const float32_t WINDOW_COHERENT_GAIN[] = {1, 0.5, 0.54, 0.42, 0.35875, 0.2155};

float32_t estimate_phase(const float32_t angle, const float32_t freq_delta,
                         const uint32_t n_pts) {
  return angle - PI * freq_delta * (n_pts - 1) / n_pts;
}

float32_t cmplx_mag_simple(const float32_t real, const float32_t img) {
  float32_t real_square = real * real;
  float32_t img_square = img * img;
  float32_t summed_square = real_square + img_square;
  float32_t mag = sqrtf(summed_square);
  return mag;
}

EstimateData parabolic_interp(const float32_t *pSrc, uint32_t peak_bin,
                              const float32_t fs_hz, const float32_t blockSize,
                              const float32_t coherent_gain) {
  uint32_t n_pts = blockSize / 2;
  float32_t y_minus1 =
      cmplx_mag_simple(pSrc[2 * (peak_bin - 1)], pSrc[2 * (peak_bin - 1) + 1]);
  float32_t y_0 = cmplx_mag_simple(pSrc[2 * peak_bin], pSrc[2 * peak_bin + 1]);
  float32_t y_plus1 =
      cmplx_mag_simple(pSrc[2 * (peak_bin + 1)], pSrc[2 * (peak_bin + 1) + 1]);
  float32_t freq_delta =
      0.5 * (y_minus1 - y_plus1) / (y_minus1 - 2 * y_0 + y_plus1);
  float32_t a = (y_minus1 + y_plus1 - 2 * y_0);
  float32_t y_peak = y_0 - a * (freq_delta * freq_delta);
  if (peak_bin > n_pts / 2)
    peak_bin = (n_pts / 2) - peak_bin;
  float32_t freq_estimated = (peak_bin + freq_delta) * (fs_hz / n_pts);
  float32_t amplitude_estimated = y_peak / (n_pts * coherent_gain);
  float32_t angle_ori;
  arm_atan2_f32(pSrc[2 * peak_bin + 1], pSrc[2 * peak_bin], &angle_ori);
  float32_t phase_estimated = estimate_phase(angle_ori, freq_delta, n_pts);
  EstimateData estimate_data;
  estimate_data.freq_estimated = freq_estimated;
  estimate_data.amplitude_estimated = amplitude_estimated;
  estimate_data.phase_estimated = phase_estimated;
  return estimate_data;
}

float32_t delta_rect_candan(const float32_t *pSrc, uint32_t peak_bin) {
  float32_t y_minus1[2], y_0[2], y_plus1[2];
  float32_t result_cmplx[2];
  y_minus1[0] = pSrc[2 * (peak_bin - 1)];
  y_minus1[1] = pSrc[2 * (peak_bin - 1) + 1];
  y_0[0] = pSrc[2 * peak_bin];
  y_0[1] = pSrc[2 * peak_bin + 1];
  y_plus1[0] = pSrc[2 * (peak_bin + 1)];
  y_plus1[1] = pSrc[2 * (peak_bin + 1) + 1];
  float32_t numerator[2], denominator[2];
  numerator[0] = y_minus1[0] - y_plus1[0];
  numerator[1] = y_minus1[1] - y_plus1[0];
  denominator[0] = 2 * y_0[0] - y_minus1[0] - y_plus1[0];
  denominator[1] = 2 * y_0[1] - y_minus1[1] - y_plus1[1];
  denominator[1] *= -1; // 求共轭
  float32_t scale =
      denominator[0] * denominator[0] - denominator[1] * denominator[1];
  arm_cmplx_mult_cmplx_f32(numerator, denominator, result_cmplx, 1);
  result_cmplx[0] /= scale;
  return result_cmplx[0];
}

typedef struct {
  float32_t r;
  float32_t alpha;
} r_and_alpha;

r_and_alpha get_r_and_alpha(const float32_t *pSrc, uint32_t peak_bin) {
  float32_t y_minus1 =
      cmplx_mag_simple(pSrc[2 * (peak_bin - 1)], pSrc[2 * (peak_bin - 1) + 1]);
  float32_t y_0 = cmplx_mag_simple(pSrc[2 * peak_bin], pSrc[2 * peak_bin + 1]);
  float32_t y_plus1 =
      cmplx_mag_simple(pSrc[2 * (peak_bin + 1)], pSrc[2 * (peak_bin + 1) + 1]);
  float32_t r;
  if (y_plus1 >= y_minus1) {
    r = 1.0f;
  } else {
    r = -1.0f;
  }
  float32_t alpha;
  if (r == 1.0f) {
    alpha = y_plus1 / y_0;
  } else if (r == -1.0f) {
    alpha = y_minus1 / y_0;
  } else {
    alpha = -114514.0f;
  }
  r_and_alpha out;
  out.r = r;
  out.alpha = alpha;
  return out;
}

float32_t delta_rect_rife(const float32_t r, const float32_t alpha) {
  return r * alpha / (1 + alpha);
}

float32_t delta_hann_granke(const float32_t r, const float32_t alpha) {
  return r * (2 * alpha - 1) / (alpha + 1);
}

float32_t delta_hamming_offelli(const float32_t r, const float32_t alpha) {
  return r * (alpha - 0.54) / (0.46 * alpha + 0.54);
}

float32_t delta_blackman_andria(const float32_t r, const float32_t alpha) {
  return r * (3 * alpha - 1.5) / (alpha + 1.5);
}

float32_t delta_blackmanharris_agrez(const float32_t r, const float32_t alpha) {
  return r * (4 * alpha - 2) / (alpha + 2);
}

float32_t delta_flattop_andria(const float32_t r, const float32_t alpha) {
  return r * (5.5 * alpha - 4.5) / (alpha + 1);
}

float32_t _sinc_delta(const float32_t delta) {
  return PI * delta / arm_sin_f32(PI * delta);
}

float32_t _m_plus_n_delta_square(const float32_t m, const float32_t n,
                                 const float32_t delta) {
  return m + n * delta * delta;
}

float32_t amp_rect(const float32_t x_k0, const float32_t freq_delta) {
  return x_k0 * _sinc_delta(freq_delta);
}

float32_t amp_hann(const float32_t x_k0, const float32_t freq_delta) {
  return x_k0 * _sinc_delta(freq_delta) *
         _m_plus_n_delta_square(1, -1, freq_delta);
}

float32_t amp_hamming(const float32_t x_k0, const float32_t freq_delta) {
  return x_k0 * _sinc_delta(freq_delta) *
         (_m_plus_n_delta_square(1, -1, freq_delta) /
          _m_plus_n_delta_square(1, -0.148, freq_delta));
}

float32_t amp_blackman(const float32_t x_k0, const float32_t freq_delta) {
  return x_k0 * _sinc_delta(freq_delta) *
         (_m_plus_n_delta_square(1, -1, freq_delta) *
          _m_plus_n_delta_square(4, -1, freq_delta) /
          _m_plus_n_delta_square(4, -1.69, freq_delta));
}

float32_t amp_blackmanharris(const float32_t x_k0, const float32_t freq_delta) {
  return x_k0 * _sinc_delta(freq_delta) *
         (_m_plus_n_delta_square(1, -1, freq_delta) *
          _m_plus_n_delta_square(1, -0.25, freq_delta) *
          _m_plus_n_delta_square(1, -0.111, freq_delta));
}

float32_t amp_flattop(const float32_t x_k0, const float32_t freq_delta) {
  float32_t freq_delta_2 = freq_delta * freq_delta;
  float32_t freq_delta_4 = freq_delta_2 * freq_delta_2;
  return x_k0 * (1 + 0.02 * freq_delta_2 + 0.0005 * freq_delta_4);
}

typedef float32_t (*delta_func_t)(float32_t, float32_t);
const delta_func_t _DELTA_METHODS[] = {
    delta_rect_rife,       delta_hann_granke,          delta_hamming_offelli,
    delta_blackman_andria, delta_blackmanharris_agrez, delta_flattop_andria};

const delta_func_t _AMP_METHODS[] = {amp_rect,           amp_hann,
                                     amp_hamming,        amp_blackman,
                                     amp_blackmanharris, amp_flattop};

EstimateData estimate_freq_amplitude_phase(const float32_t *pSrc,
                                           const uint32_t blockSize,
                                           const uint32_t peak_bin,
                                           const uint32_t fs_hz,
                                           const uint32_t window_type,
                                           const uint32_t algorithm_type) {
  uint32_t n_pts = blockSize / 2;
  EstimateData out;
  if ((peak_bin == 0) || (peak_bin > n_pts - 2)) {
    out.amplitude_estimated = -114514.0f;
    out.phase_estimated = -114514.0f;
    return out;
  }
  float32_t coherent_gain = WINDOW_COHERENT_GAIN[window_type];
  float32_t x_k0 = cmplx_mag_simple(pSrc[2 * peak_bin], pSrc[2 * peak_bin + 1]);
  float32_t angle_ori, freq_delta, amp_estimated, phase_estimated;
  r_and_alpha ralpha;
  arm_atan2_f32(pSrc[2 * peak_bin + 1], pSrc[2 * peak_bin], &angle_ori);
  switch (algorithm_type) {
  case PARABOLIC:
    return parabolic_interp(pSrc, peak_bin, fs_hz, blockSize, coherent_gain);
  case CANDAN:
    if (window_type != RECT) {
      out.amplitude_estimated = -114514.0f;
      out.phase_estimated = -114514.0f;
      return out;
    }
    freq_delta = delta_rect_candan(pSrc, peak_bin);
    break;
  case AUTO:
    switch (window_type) {
    case RECT:
      freq_delta = delta_rect_candan(pSrc, peak_bin);
      break;
    case HANN:
      ralpha = get_r_and_alpha(pSrc, peak_bin);
      float32_t r = ralpha.r;
      float32_t alpha = ralpha.alpha;
      freq_delta = delta_hann_granke(r, alpha);
      break;
    default:
      return parabolic_interp(pSrc, peak_bin, fs_hz, blockSize, coherent_gain);
    }
  default:
    if (algorithm_type == AUTO)
      break; // 特殊情况排除
    ralpha = get_r_and_alpha(pSrc, peak_bin);
    float32_t r = ralpha.r;
    float32_t alpha = ralpha.alpha;
    freq_delta = _DELTA_METHODS[algorithm_type](r, alpha);
    break;
  }
  amp_estimated = _AMP_METHODS[window_type](x_k0, freq_delta);
  amp_estimated /= n_pts * coherent_gain;
  phase_estimated = estimate_phase(angle_ori, freq_delta, n_pts);
  out.freq_estimated = (peak_bin + freq_delta) * (fs_hz / (float32_t)n_pts);
  out.amplitude_estimated = amp_estimated;
  out.phase_estimated = phase_estimated;
  return out;
}
